src/main/java/com/iexec/blockchain/broker/BrokerService.java

```diff
@@ -115,41 +111,64 @@ private Optional<String> fireMatchOrders(BrokerOrder brokerOrder) {
         return Optional.empty();
     }
 
-
-    private boolean hasRequesterAcceptedPrices(
+    boolean hasRequesterAcceptedPrices(
             RequestOrder requestOrder,
             BigInteger appPrice,
             BigInteger workerpoolPrice,
-            BigInteger datasetPrice
+            BigInteger datasetPrice,
+            boolean withDataset
     ) {
-        if (requestOrder == null || requestOrder.getWorkerpoolmaxprice() == null
-                || requestOrder.getAppmaxprice() == null
-                || appPrice == null || workerpoolPrice == null) {
+        try {
+            boolean isAppPriceAccepted = requestOrder.getAppmaxprice().longValue() >= appPrice.longValue();
+            boolean isPoolPriceAccepted = requestOrder.getWorkerpoolmaxprice().longValue() >= workerpoolPrice.longValue();
+            boolean isAccepted = isAppPriceAccepted && isPoolPriceAccepted;
+            String messageDetails = MessageFormat.format("[isAppPriceAccepted:{0}, isPoolPriceAccepted:{1}]",
+                    isAppPriceAccepted, isPoolPriceAccepted);
+            if (withDataset) {
+                boolean isDatasetPriceAccepted = requestOrder.getDatasetmaxprice().longValue() >= datasetPrice.longValue();
+                isAccepted = isAccepted && isDatasetPriceAccepted;
+                messageDetails = MessageFormat.format("[isAppPriceAccepted:{0}, isPoolPriceAccepted:{1}, isDatasetPriceAccepted:{2}]",
+                        isAppPriceAccepted, isPoolPriceAccepted, isDatasetPriceAccepted
+                );
+            }
+            if (!isAccepted) {
+                log.error("Prices not accepted (too expensive) {}", messageDetails);
+            }
+            return isAccepted;
+        } catch (NullPointerException e) {
             log.error("Failed to check hasRequesterAcceptedPrices (null requestOrder)");
             return false;
         }
```

Comment:
The problem here is the complexity of the null checks to perform on requestorder and its content depending on the request type.
Previously, the check was done the way you suggest it and there was a big miss on dataset price.
By catching NPEs globally, we guarantee that we won't miss any NPE without having to check every single combination.

Reply:
Still, I agree with you that this is not my dream implementation.
What remains to be done (will require several steps) is:
1. Add sane default and forbid null values in some fields of those classes https://github.com/iExecBlockchainComputing/iexec-common/tree/develop/src/main/java/com/iexec/common/sdk/order/payload
2. Rework the implementation here at this time

But I would prefer to keep the current implementation as a relatively sane/protected algorithm avoiding weird behaviors and uncaught NPEs due to the current data model.

Reply:
I still don't like this approach. IMHO we should check each value independently. If you don't want to bother with a lot of `if` blocks, you maybe could do the following:
```
        try {
            Objects.requireNonNull(obj1, "Expecting obj1 not to be null");
            Objects.requireNonNull(obj2, "Expecting obj2 not to be null");
            Objects.requireNonNull(obj3, "Expecting obj3 not to be null");
            [...]
        } catch (NullPointerException e) {
            log.error("Failed to check hasRequesterAcceptedPrices (null requestOrder)", e);
            return false;
        }
```
This would allow to fail fast instead of computing while we know it will fail later.

Reply:
BTW, the way you test whether values are null here does not really tell us which value is null. E.g: l123, how would we know which value is null? There could be 2 or 3 NPE thrown on this single line. You don't even log the exception, so we don't have any trace of where it failed. Maybe consider add the exception to the logs?

Reply:
Thanks! I agree with you on the note. You should maybe add a TODO or submit an issue on GitHub so we don't forget it.